### DISCUSSION TITLE
NextJS v12.3.1 CSP blocks use of eval in prod

### DISCUSSION THREAD

**Discussion by michaelcais** [vercel/next.js#41072] [Help] [answered]
I've set up a strict CSP policy for script-src and on one section of my app has triggered a CSP error on production. I tried to trace the corresponding error by enabling source map and it pointed to a file called util.js that is not part of my own script. I compared this section that has the CSP error to another section of my app with similar code but without the CSP error and I couldn't find anything in particular or difference in code or plugin library that would cause this eval issue on one and not the other. I do not have any special webpack config on my next.config.js file.

```
const headers = require('./headers');

const withBundleAnalyzer = require('@next/bundle-analyzer')({
  enabled: process.env.ANALYZE === 'true',
  openAnalyzer: false
});

const nextConfig = {
  reactStrictMode: true,
  productionBrowserSourceMaps: true,
  swcMinify: true,
  images: {
    domains: [
      'images.unsplash.com'
    ]
  },
  poweredByHeader: false,
  async headers() {
    return [
      {
        source: '/:path*',
        headers
      }
    ];
  }
};

module.exports = withBundleAnalyzer(nextConfig);
```

The eval error seems to point to this particular code at line 690 from util.js

`return Function('return function*() {}')();`

The section of the app with the CSP error seems to work just fine but I'm concerned to see any errors with issues that may not be immediately apparent.
[util.js.zip](https://github.com/vercel/next.js/files/9686984/util.js.zip)

**Comment by michaelcais** [comment, marked as answer]
I finally found what was triggering the eval issue. It was actually one of my client side imports that has server side script with some dependencies on external library. I moved the function that I needed from the imported file with server side script to my own utils file and that fixed my CSP eval error.

**Comment by matt-vendia** [comment]
@michaelcais Can you share any more info on the file/lib that was the root cause? Facing the same issue. Thanks!

**Reply by michaelcais** [reply]
@matt-vendia I'm not sure if it helps your issue, but I had a higher order function for cookie that was supposed to be only for server-side for sealing and parsing cookie that I incorrectly imported to a client-side module. This higher order function accepts the context from getServerSideProps and returns a function to handle cookies.

**Reply by matt-vendia** [reply]
Interesting. Well, at least it's another thread to tug at. I'll update if I uncover any helpful info. Thanks!

**Comment by mauritsl** [comment]
This code is in `node_modules/next/dist/compiled/util/util.js` and unminified it looks like this:

```
a = (function () {
  if (!o) return !1;
  try {
    return Function('return function*() {}')();
  } catch (e) {}
})(),
```

That tests for the availability of generator functions (all major browsers after IE11), but triggers a CSP eval error. It was introduced in version 12.0.8, perhaps by commit 3667eba385. This issue is still present in 13.1.2. It's not triggered on all pages, so likely depends on libraries used.

Can't we drop the regenerator polyfill (or at least give the option to do so?). IE11 is not in the list of supported browsers for Next.
